shaping/wrapping.go

```diff
@@ -0,0 +1,243 @@
+package shaping
+
+import (
+	"github.com/go-text/typesetting/di"
+	"github.com/npillmayer/uax/segment"
+	"github.com/npillmayer/uax/uax14"
+)
+
+// mapRunesToClusterIndices returns a slice. Each index within that slice corresponds
+// to an index within the runes input slice. The value stored at that index is the
+// index of the glyph at the start of the corresponding glyph cluster shaped by
+// harfbuzz.
+func mapRunesToClusterIndices(runes []rune, glyphs []Glyph) []int {
+	mapping := make([]int, len(runes))
+	glyphCursor := 0
+	if len(runes) == 0 {
+		return nil
+	}
+	// If the final cluster values are lower than the starting ones,
+	// the text is RTL.
+	rtl := len(glyphs) > 0 && glyphs[len(glyphs)-1].ClusterIndex < glyphs[0].ClusterIndex
+	if rtl {
+		glyphCursor = len(glyphs) - 1
+	}
+	for i := range runes {
+		for glyphCursor >= 0 && glyphCursor < len(glyphs) &&
+			((rtl && glyphs[glyphCursor].ClusterIndex <= i) ||
+				(!rtl && glyphs[glyphCursor].ClusterIndex < i)) {
+			if rtl {
+				glyphCursor--
+			} else {
+				glyphCursor++
+			}
+		}
+		if rtl {
+			glyphCursor++
+		} else if (glyphCursor >= 0 && glyphCursor < len(glyphs) &&
+			glyphs[glyphCursor].ClusterIndex > i) ||
+			(glyphCursor == len(glyphs) && len(glyphs) > 1) {
+			glyphCursor--
+			targetClusterIndex := glyphs[glyphCursor].ClusterIndex
+			for glyphCursor-1 >= 0 && glyphs[glyphCursor-1].ClusterIndex == targetClusterIndex {
+				glyphCursor--
+			}
+		}
+		if glyphCursor < 0 {
+			glyphCursor = 0
+		} else if glyphCursor >= len(glyphs) {
+			glyphCursor = len(glyphs) - 1
+		}
+		mapping[i] = glyphCursor
+	}
+	return mapping
+}
+
+// inclusiveGlyphRange returns the inclusive range of runes and glyphs matching
+// the provided start and breakAfter rune positions.
+// runeToGlyph must be a valid mapping from the rune representation to the
+// glyph reprsentation produced by mapRunesToClusterIndices.
+// numGlyphs is the number of glyphs in the output representing the runes
+// under consideration.
+func inclusiveGlyphRange(start, breakAfter int, runeToGlyph []int, numGlyphs int) (glyphStart, glyphEnd int) {
+	rtl := runeToGlyph[len(runeToGlyph)-1] < runeToGlyph[0]
+	runeStart := start
+	runeEnd := breakAfter
```

Comment:
Are the new variables really needed ?

Reply:
Nope, these are another artifact of an older version of the code.